### PR TITLE
ARROW-6071: [C++] Generic binary-to-binary casts

### PR DESCRIPTION
**Pull request by pitrou**
Implement all flavours of binary-to-binary casting, except for fixed-size binary.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_cast_string.cc [hunk 1]
```diff
@@ -92,12 +93,74 @@ struct Utf8Validator {
...
+
+// Cast same-width offsets (no-op)
+template <typename I, typename O>
+void CastBinaryToBinaryOffsets(enable_if_t<has_same_width<I, O>(), KernelContext*> ctx,
+                               const ArrayData& input, ArrayData* output) {}
+
+// Upcast offsets
+template <typename I, typename O>
+void CastBinaryToBinaryOffsets(enable_if_t<has_smaller_width<I, O>(), KernelContext*> ctx,
+                               const ArrayData& input, ArrayData* output) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_cast_string.cc:113]:
```suggestion
template <typename I, typename O>
void CastBinaryToBinaryOffsets(enable_if_t<has_smaller_width<I, O>() || has_smaller_width<O, I>(), KernelContext*> ctx,
                               const ArrayData& input, ArrayData* output) {
  if (has_smaller_width<I, O>()) {
    //
  } else {
    //
  }
```

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_cast_string.cc:113]: Lack of `constexpr if` risks producing compile errors.

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_string.cc:113]: If constexpr can be emulated using `std::conditional<has_smaller_width<I, O>(), NarrowerImpl, WiderImpl>::template Exec<I, O>(ctx, input, output)`

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_cast_string.cc:113]: Horror.

**Review by bkietz — APPROVED** [review]

**Pull request closed without merging**